### PR TITLE
fix: add max_history_tokens to prevent session overflow (fixes #5011)

### LINKED ISSUE #5011: [Bug] Session Summary surpass the model max length
### Description

Image
session_runs to long , i didn't find argument to control this 

### Steps to Reproduce

Image
enable this parameter 

### Agent Configuration (if applicable)

_No response_

### Expected Behavior

a parameter in Agent BaseModel to control this behavior about input summary length, to avoid exceed max len  

### Actual Behavior

can't find such params

### Screenshots or Logs (if applicable)

_No response_

### Environment

```markdown
Agno version latest 2025.10.13
```

### Possible Solutions (optional)

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by giulio-leone**
## Summary

Fixes #5011 — When `add_history_to_context` is enabled, session history can grow unboundedly and exceed the model's context window, causing API errors.

## Changes

- Added `max_history_tokens` parameter to `Agent` (optional, defaults to `None` for backward compatibility)
- Added `truncate_history_by_tokens()` utility that drops oldest history messages to fit within a token budget
- Applied truncation in both sync (`get_run_messages`) and async (`aget_run_messages`) paths
- Token estimation uses ~4 characters per token heuristic

## Usage

```python
agent = Agent(
    add_history_to_context=True,
    max_history_tokens=4000,  # keep history within 4k tokens
)
```

## Tests

- 13 unit tests covering token estimation, truncation logic, edge cases, and Agent parameter acceptance
- All 276 existing agent unit tests continue to pass

### CODE REVIEW

File: libs/agno/agno/utils/message.py [hunk 1]
```diff
@@ -65,6 +65,54 @@ def filter_tool_calls(messages: List[Message], max_tool_calls: int) -> None:
...
+    elif isinstance(message.content, list):
+        char_count += len(json.dumps(message.content))
+    if message.tool_calls:
+        char_count += len(json.dumps(message.tool_calls))
+    if message.role:
+        char_count += len(message.role)
+    if message.name:
+        char_count += len(message.name)
+    # ~4 characters per token
+    return max(char_count // 4, 1)
```
[the remark this answers is not in the document]
Comment (giulio-leone) [reply libs/agno/agno/utils/message.py]: Fixed in cec0a1f — changed to ceiling division `(char_count + 3) // 4` so estimates are conservative and prevent context-window overflows.

File: libs/agno/agno/utils/message.py [hunk 2]
```diff
@@ -65,6 +65,54 @@ def filter_tool_calls(messages: List[Message], max_tool_calls: int) -> None:
...
+    """Estimate the token count for a message using the ~4 chars per token heuristic."""
+    import json
+
+    char_count = 0
+    if isinstance(message.content, str):
+        char_count += len(message.content)
+    elif isinstance(message.content, list):
+        char_count += len(json.dumps(message.content))
+    if message.tool_calls:
+        char_count += len(json.dumps(message.tool_calls))
```
[the remark this answers is not in the document]
Comment (giulio-leone) [reply libs/agno/agno/utils/message.py]: Fixed in cec0a1f — extracted a `_safe_json_length()` helper that uses `json.dumps(value, default=str)` with a `try/except` fallback to `len(str(value))`. This prevents `TypeError` on non-serializable content.

File: libs/agno/tests/unit/agent/test_max_history_tokens.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        msg = Message(role="user", content="Hello world")
+        tokens = _estimate_message_tokens(msg)
+        # "Hello world" = 11 chars + "user" = 4 chars => 15 // 4 = 3
+        assert tokens >= 1
+
+    def test_empty_content(self):
+        msg = Message(role="user", content="")
+        tokens = _estimate_message_tokens(msg)
+        assert tokens >= 1
+
```
[the remark this answers is not in the document]
Comment (giulio-leone) [reply libs/agno/tests/unit/agent/test_max_history_tokens.py]: Fixed in cec0a1f — tests now assert exact expected token counts for string content, verify reasonable ranges for list content, check meaningful differences for tool_calls, and include a test for non-serializable content.

File: libs/agno/agno/agent/_messages.py [hunk 4]
```diff
@@ -1459,6 +1463,10 @@ async def aget_run_messages(
             if agent.max_tool_calls_from_history is not None:
                 filter_tool_calls(history_copy, agent.max_tool_calls_from_history)
 
+            # Truncate history by token budget if set
+            if agent.max_history_tokens is not None:
+                truncate_history_by_tokens(history_copy, agent.max_history_tokens)
+
```
[the remark this answers is not in the document]
Comment (giulio-leone) [reply libs/agno/agno/agent/_messages.py:1490]: Fixed in cec0a1f — the fix is in `truncate_history_by_tokens()` which is shared by both sync and async paths, so both are now covered by the same leading-tool-message cleanup logic.

**Review by giulio-leone — COMMENTED** [review]
Submitting pending review to unblock comment replies.

File: libs/agno/agno/agent/agent.py [hunk 5]
```diff
@@ -131,6 +131,10 @@ class Agent:
     num_history_messages: Optional[int] = None
     # Maximum number of tool calls to include from history (None = no limit)
     max_tool_calls_from_history: Optional[int] = None
+    # Maximum number of tokens from history to include in context (None = no limit).
+    # When set, truncates the oldest history messages to fit within this token budget.
+    # Uses approximate token counting (1 token ≈ 4 characters).
+    max_history_tokens: Optional[int] = None
```
[the remark this answers is not in the document]
Comment (giulio-leone) [reply libs/agno/agno/agent/agent.py:139]: Fixed in 3fa28fee6 — added `max_history_tokens` to both `to_dict()` and `from_dict()` in `_storage.py`, following the same pattern as `max_tool_calls_from_history`. A round-trip test (`test_max_history_tokens_round_trip`) confirms the value survives serialization/deserialization.

File: libs/agno/agno/utils/message.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (giulio-leone) [reply libs/agno/agno/utils/message.py]: Fixed — changed from floor division to conservative ceiling division: `(char_count + 3) // 4` which is equivalent to `math.ceil(char_count / 4)`. This ensures token estimates round up, preventing more history from leaking through than the budget intends.

File: libs/agno/tests/unit/agent/test_max_history_tokens.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (giulio-leone) [reply libs/agno/tests/unit/agent/test_max_history_tokens.py:19]: Fixed — tests now include exact assertions (e.g. `assert tokens == 4`), range checks (`5 <= tokens <= 20`), and comparative assertions (`tokens_with > tokens_without` with a minimum delta of 5). A non-serializable content test and a storage round-trip test were also added.

File: libs/agno/agno/agent/_messages.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (giulio-leone) [reply libs/agno/agno/agent/_messages.py:1490]: Fixed — the async path at line 1468 calls the same `truncate_history_by_tokens()` utility function which already handles dropping leading tool messages. Both sync and async code paths share this logic, so the fix applies to both.

File: libs/agno/agno/utils/message.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (giulio-leone) [reply libs/agno/agno/utils/message.py]: Fixed — introduced `_safe_json_length()` which wraps `json.dumps(value, default=str)` in a try/except, falling back to `len(str(value))` if serialization fails. Both `message.content` (list) and `message.tool_calls` now go through this safe path. A test with a non-serializable `Custom` object confirms no exception is raised.

File: libs/agno/agno/agent/_messages.py [hunk 6]
```diff
@@ -1257,6 +1257,10 @@ def get_run_messages(
...
```
[the remark this answers is not in the document]
Comment (giulio-leone) [reply libs/agno/agno/agent/_messages.py]: Fixed — `truncate_history_by_tokens()` now drops leading tool messages after truncation via a while-loop that pops orphaned tool messages from the front. This is in the shared utility function called by both sync and async paths. A dedicated test (`test_drops_leading_tool_messages`) validates this behavior.

**Comment by giulio-leone** [comment]
Hi! Gentle ping — this PR is rebased, CI passes, and ready for review. Happy to address any feedback. Thanks!

**Comment by giulio-leone** [comment]
Hi! Gentle ping — this PR is rebased, CI passes, and ready for review. Happy to address any feedback. Thanks!

**Comment by giulio-leone** [comment]
Friendly ping — CI is green, tests pass, rebased on latest. Ready for review whenever convenient. Happy to address any feedback. 🙏

**Comment by giulio-leone** [comment]
Hi — all 6 review threads resolved. Ready for review!
